### ISSUE TITLE
instanceof behaves wrongly with empty object and primitive type union.

### ISSUE THREAD

**Issue by mikelehen** [Microsoft/TypeScript#15584]
This may be a dupe of #14426 but the manifestation is a bit different.

**TypeScript Version:**  2.3.2

**Code**

```ts
// Type alias to represent un-typed code from external JavaScript.
type AnyJs = null | undefined | boolean | number | string | { };

abstract class OpaqueHandle {
}

class IdHandle extends OpaqueHandle {
  constructor(readonly id: number) { super(); }
}

function process(input: AnyJs) {
  if (input instanceof OpaqueHandle) {
    // Following line produces:  TS2358: The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter.
    if (input instanceof IdHandle) { // input is wrongly inferred to string|number|boolean
      console.log(input.id);
    }
  }
}
```

**Expected behavior:** No errors.

**Actual behavior:** Error:
```
test.ts(14,9): error TS2358: The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter.
```

I can fix the problem by wrapping the function body in a `if (typeof(input) === 'object') { ... }` conditional.

**Comment by mhegazy** [comment]
Just define `input` as `any`.

**Comment by mikelehen** [comment]
Well, I'm explicitly trying to avoid "any" since in addition to being a supertype of all other types, it's also a subtype, so you lose all type-safety (e.g. the compiler would let me do `input.id` without first doing the proper instanceof checks).  

And the fact that using `any` avoids the error doesn't change the fact that the code should work fine as-is. :-)

**Comment by kitsonk** [comment]
It is a supertype, but the `instanceof` operator does narrow the type:

```ts
abstract class OpaqueHandle {
}

class IdHandle extends OpaqueHandle {
  constructor(readonly id: number) { super(); }
}

function process(input: any) {
  if (input instanceof OpaqueHandle) {
    if (input instanceof IdHandle) {
      console.log(input.id);
    }
  }
}
```

This is mainly because `instaceof` is weird, in that the rest of ECMAScript is largely a structural type system, but `instanceof` is nominal... So TypeScript bridges that.  TypeScript will narrow `any` in situations where it is confident it can be narrowed.  For example:

```ts
function foo(input: any) {
  if (typeof input === 'string') {
    input; // type is now 'string'
  }
}
```

But if you do insist on re-inventing the wheel, [the following works fine](http://www.typescriptlang.org/play/#src=type%20AnyJs%20%3D%20null%20%7C%20undefined%20%7C%20boolean%20%7C%20number%20%7C%20string%20%7C%20object%3B%0A%0Aabstract%20class%20OpaqueHandle%20%7B%0A%7D%0A%0Aclass%20IdHandle%20extends%20OpaqueHandle%20%7B%0A%20%20constructor(readonly%20id%3A%20number)%20%7B%20super()%3B%20%7D%0A%7D%0A%0Afunction%20process(input%3A%20AnyJs)%20%7B%0A%20%20if%20(input%20instanceof%20OpaqueHandle)%20%7B%0A%20%20%20%20if%20(input%20instanceof%20IdHandle)%20%7B%0A%20%20%20%20%20%20console.log(input.id)%3B%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A):

```ts
type AnyJs = null | undefined | boolean | number | string | object;

abstract class OpaqueHandle {
}

class IdHandle extends OpaqueHandle {
  constructor(readonly id: number) { super(); }
}

function process(input: AnyJs) {
  if (input instanceof OpaqueHandle) {
    if (input instanceof IdHandle) {
      console.log(input.id);
    }
  }
}
```

Because `{ } !== object` in the type system.  In versions of TypeScript prior to 2.2 (IIRC) `object` was a compiler concept but not exposed in the language, where `{ }` is a bit of a strange type that is needed to deal with all sorts of strange oddities with type assignment.  It is essentially `any` in a lot of ways as a type, but not quite.

**Comment by mikelehen** [comment]
Ah!  I didn't know I could use `object`.  It does appear that will solve my issue in this case.  I still don't understand the behavior with `{}`  (it seems like a bug), but I'm definitely unblocked.  Thanks!

FWIW, I don't think I'm "reinventing the wheel."  I'm aware that you /can/ narrow `any` to a more specific type, but the problem is that you don't /have/ to.  The compiler would let me reference `input.id` without any instanceof checks, which would be a bug.  By using AnyJs, the compiler still requires me to have the right instanceof / typeof checks, which helps ensure I properly handle all the cases without bugs.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
